applications/callflow/src/cf_blacklist.erl

```diff
@@ -0,0 +1,175 @@
+
+-spec compile_patterns(kz_term:ne_binary(), kz_term:api_ne_binary(), patterns_view_items(), patterns()) -> lookup_ret_patterns().
+compile_patterns(AccountId, OwnerId, [], Patterns) ->
+    cache_patterns(AccountId, OwnerId, Patterns);
+compile_patterns(AccountId, OwnerId, [{Regex, JObj} | Rest], Acc) ->
+    Action = kz_json:get_ne_binary_value(<<"action">>, JObj, <<"block">>),
+    CallerName = kz_json:get_ne_binary_value(<<"name">>, JObj),
+    case re:compile(Regex) of
+        {'ok', {'re_pattern', Groups, _, _, _} = MP}
+          when Groups =:= 0 ->
+            Pattern = #pattern{action=Action, caller_name=CallerName, regex=MP, has_groups='false'},
+            compile_patterns(AccountId, OwnerId, Rest, [Pattern | Acc]);
+        {'ok', MP} ->
+            {'namelist', Names} = re:inspect(MP, 'namelist'),
+            Pattern = #pattern{action=Action, caller_name=CallerName, regex=MP, names=Names, has_groups='true'},
+            compile_patterns(AccountId, OwnerId, Rest, [Pattern | Acc]);
+        _Err ->
+            lager:debug("unexpected result compiling regular expression : ~p", [_Err]),
+            compile_patterns(AccountId, OwnerId, Rest, Acc)
+    end.
+
+-spec cache_patterns(kz_term:ne_binary(), kz_term:api_ne_binary(), patterns()) -> lookup_ret_patterns().
+cache_patterns(AccountId, OwnerId, Patterns) ->
+    AccountDb = kz_util:format_account_db(AccountId),
+    CacheOptions = [{'origin', [{'db', AccountDb, <<"blacklist">>}]}
+                   ,{'expires', ?MILLISECONDS_IN_HOUR}
+                   ],
+    kz_cache:store_local(?CACHE_NAME, ?BL_PATTERNS_CACHE_KEY(AccountId, OwnerId), Patterns, CacheOptions),
+    case Patterns of
+        [] -> {'error', 'not_found'};
+        _ -> {'ok', Patterns}
+    end.
+
+-spec cache_number(kz_term:ne_binary(), kz_term:api_ne_binary(), kz_term:ne_binary(), kz_term:ne_binary(), kz_term:api_ne_binary()) -> lookup_ret_number().
+cache_number(AccountId, OwnerId, Number, Action, CallerName) ->
+    AccountDb = kz_util:format_account_db(AccountId),
+    CacheOptions = [{'origin', [{'db', AccountDb, <<"blacklist">>}]}
+                   ,{'expires', ?MILLISECONDS_IN_HOUR}
+                   ],
+    kz_cache:store_local(?CACHE_NAME, ?BL_NUMBER_CACHE_KEY(AccountId, OwnerId, Number), {Action, CallerName}, CacheOptions),
+    case Action of
+        'undefined' -> {'error', 'not_found'};
+        _ -> {'ok', Action, CallerName, Number =:= ?NO_MATCH_BL}
+    end.
+
+-spec maybe_use_nomatch(kz_term:ne_binary(), kz_term:ne_binary(), kz_term:api_ne_binary(), kz_term:ne_binary()) -> lookup_ret_number().
+maybe_use_nomatch(Number, AccountId, OwnerId, Strategy) ->
+    case knm_converters:is_reconcilable(Number, AccountId) of
+        'true' -> try_nomatch(Number, AccountId, OwnerId, Strategy);
+        'false' ->
+            lager:info("can't use no_match: number not all digits: ~s", [Number]),
+            cache_number(AccountId, OwnerId, Number, 'undefined', 'undefined')
```

Comment:
This breaks the contract with `cache_number/5`. Looks like `Action` is expected to be `ne_binary()` in the spec.